### ISSUE TITLE
the unit is "stuck" in the river ford

### ISSUE THREAD

**Issue by MrZorG33** [We-the-People-civ4col-mod/Mod#449]
starting a new game with the latest update of files, I saw that the units can again "get stuck" in the river ford. this has not happened to my units yet. but the Spanish settler has been standing motionless in the river ford for several turns.

![юнит застрял в броде01](https://user-images.githubusercontent.com/61798883/116873435-59c5d380-ac20-11eb-9c15-dee0d2f06523.jpg)

[large river.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/6414654/large.river.zip)

**Comment by Nightinggale** [comment]
Most likely not stuck as such, but more like an AI issue. My guess is for optimization purposes the AI only checks the same area when looking for a place to build a new colony. Since it's standing in the river, it checks the river and ends turn because it failed to find a valid location.

**Comment by MrZorG33** [comment]
then, in theory, he should move on, but he stands still.

**Comment by raystuttgart** [comment]
Ah guys, this is a settler !
I have tried to adapt settler logic for Large Rivers.
May never have considered though that a Settler actually is ON TOP of a Large River.

Will need to check.

**Comment by raystuttgart** [comment]
Edit, could not find a reason in my logic for this to fail.
Area check seems plausible though.

Any help is appreciated.
Will however also analyze the next days.

**Until the weekend** I will be pretty busy and wasted though ...

**Comment by MrZorG33** [comment]
new data!
the Spaniards have declared war on me, but their troops also do not go beyond the river ford.

![image](https://user-images.githubusercontent.com/61798883/117054344-ac87b400-ad22-11eb-8a01-9ecbb417c111.png)
[war.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/6423312/war.zip)

**Comment by raystuttgart** [comment]
@Nightinggale 
Could you maybe check what is happening with the Military Unit here?
I have no clue how I could fix it.

What I currently assume is that it might be related to **"Amphibious Landing"** or some other part of the Logic of Combat Units.
(Similar to the logic when Military Units unboard a Ship.)

**Comment by Nightinggale** [comment]
I can't load the savegame and I can't recreate the setup where a unit is stuck on a river.

**Comment by MrZorG33** [comment]
What is the problem with loading the saved game?

**Comment by Nightinggale** [comment]
Crash on load. I'm referring to the soldier one.

**Comment by MrZorG33** [comment]
most likely this is due to the new changes that have been made to the mod. I already had this problem. unfortunately I've already updated my files and most likely won't be able to load the saved game either.

**Comment by Nightinggale** [comment]
Ok, I got it to load using an old version. Now there are 3 issues to investigate:

- why does the AI decide that ACTIVITY_HOLD is the best option on that plot?
- why can't the savegame be loaded even though (to my knowledge) no savegame breaking changes have been made?
- checkPower assertion failure (AI can't determine power of players correctly)

**Comment by MrZorG33** [comment]
Note again: the Spanish troops cannot get out of the river ford into my territory, but they can into their territory.

**Comment by Nightinggale** [comment]
The fix should work this time.

**Comment by raystuttgart** [comment]
Awesome !!! 
How did you fix it?

**Comment by Nightinggale** [comment]
Fix is in the git log 437b652a0862d5706515766fd49f0a2461e32c3c

**Comment by raystuttgart** [comment]
Ah ok, sou you really added the parameter "Domain" to getArea calls !
And then you check also in getArea for Large Rives

I once thought about it but I was afraid to mess up performance ...
Does it not impact performance?

**Comment by Nightinggale** [comment]
I didn't measure performance, but I don't think it will. If you check the two new functions they are vanilla code except if it's DOMAIN_LAND and TERRAIN_LARGE_RIVERS. That combo is rare enough to not trigger often. Secondly when it triggers, it will loop 9 plots, but it stops at the first land plot, which is usually within the first 3, and not unlikely the first. Performance is good enough, but can be made better with #464. That will however take quite a while to code and it's not urgent.

**Comment by MrZorG33** [comment]
while testing the "new hope" assembly, after some time I ran into an old bug (unfortunately I could not find a similar issue. perhaps the conversation was in a working chat) - automated transports "stuck" in the river ford. moreover, for some time they successfully used it, but after a while I noticed their absence (and the lack of goods in warehouses, hehe), I started checking the map and found them all in one place.
I assume that the root of the problem is the same as in the initial issue. so i posted it here.
![image](https://user-images.githubusercontent.com/61798883/200092545-2e64deb6-bc37-44c9-b61c-df9bf7e1eb30.png)

update November 3rd 15:45 (utc+3) (before Ray added new music))
[NWGiganticMy.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/9942620/NWGiganticMy.zip)

if it fails to open this save, I will update the mod and start the test again.

**Comment by raystuttgart** [comment]
@MrZorG33 Did you compile a new DLL?

**Comment by raystuttgart** [comment]
[@nightinggale](https://colonization-mod.slack.com/team/U1YBGQHLY)
 I think I found the bug for the "get stuck on River Ford" of Automated Transports !!!
It is in here:
bool CvSelectionGroupAI::AI_tradeRoutes()
-----------
In particular I think it is this section:
			if ((pSourceCity != NULL) && ((domainType != DOMAIN_SEA) || (pSourceWaterArea != NULL)))
			{
				int iSourceArea = (domainType == DOMAIN_SEA) ? pSourceWaterArea->getID() : pSourceCity->getArea();
				if (domainType == DOMAIN_SEA ? plot()->isAdjacentToArea(iSourceArea) : (iSourceArea == getArea()) ||
					domainType == DOMAIN_LAND && plot()->getTerrainType() == TERRAIN_LARGE_RIVERS && plot()->isAdjacentToArea(pSourceCity->getArea()))
				{
					if ((domainType == DOMAIN_SEA) || (pRoute->getDestinationCity() != kEurope))
					{
						processTradeRoute(pRoute, cityValues, routes, routeValues, yieldsDelivered, yieldsToUnload);
					}
				}
				// TODO: Need to path find if the area check fails like below
			}
-----------
I think this also needs to have the call with getArea(DOMAIN_LAND).
What do you think ?

**Comment by Nightinggale** [comment]
After a bunch of debugging it seems that the problem is that all destinations are either full or unreachable. Interestingly not unreachable from the plot the units are on, but rather unreachable when making a path from source colony to destination colony.

Usually transports will then hide in colonies until they have something to do. However that part of the code seems to only find colonies in the same area. The large rivers are different areas and in fact the stagecoach to the west is also stuck because it is on an "island" surrounded by large rivers, hence a different area than the colonies.

In short: the bug is that they park in the wrong location when idle rather than transports failing to do work they are supposed to do. Still not ideal as the vanilla code moves them to colonies for protection.

**Comment by MrZorG33** [comment]
well, my scout is stuck on an island surrounded by LR. he explored the territory in automatic mode.
![image](https://user-images.githubusercontent.com/61798883/208309684-529fbca5-46e2-471b-bb3a-329f49272768.png)
[qwerty.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/10254271/qwerty.zip)

if you look at the map (globe), you will see that he still has a lot of work to do. but, apparently, he was tired and there is a beautiful view))

**Comment by Nightinggale** [comment]
I have a pretty good idea of what has gone wrong here and a complete fix for the previous problem should fix this too. It's however not trivial to fix so I want to finish other tasks first.

The real issue is that the island is a different area because it's surrounded by water and the scout is looking for places to explore in the same area, hence it searches the island for places to explore or colonies to move to. There aren't any so it's just stuck. The area code still has some flaws regarding large rivers.

**Comment by devolution79** [comment]
We had a discussion a while ago (lost in the slack history paywall sadly :( ) where we brainstormed (myself and @Nightinggale) potential solutions to this and related issues. I cannot recall all the details of our exchange but core issue is that large rivers should not "break apart" land areas. If they are allowed to do so, we end up with a bunch of problems and ad-hoc workarounds that result all sort of weird bugs. My solution is to modify the land area generator so that land areas containing large rivers stay as a single area.
Update: I've implemented this and currently testing the result

**Comment by devolution79** [comment]
Just to clarify, large rivers do remain part of the "overall" water area it's just that the land areas on both sides will be in the same area. This should end the unit ai confusion since it will realize that it can simply pathfind over the nearest river ford to continue its traversal of the area. This also means that the military / native AI has been un-nerfed and they can now consider to cross large rivers in order to carry out their attack (without this fix they would have to undertake an intercontinental invasion!)

**Comment by raystuttgart** [comment]
Thanks for giving this a try to fix it. 👍 

@Nightinggale had explained me as well that the "Large Rivers not splitting Areas" would be the ideal solution he would also try.
So what you are going to implement or already have implemented sounds just perfect and logical.

**Comment by Nightinggale** [comment]
> We had a discussion a while ago (sadly lost to the slack history paywall 😢 ) where we brainstormed (myself and @Nightinggale) potential solutions to this and related issues. 

The idea was to make an array of areas in each plot instead of just one. This way each plot can hold a land and a water area. Cities and large rivers can then merge both types.

**Comment by raystuttgart** [comment]
**@devolution79** 
Yesterday you wrote that you had already implemented a solution to "Large Rivers not splitting Areas".
Are you still testing it or did you decide to do some redesign after the discussion with @Nightinggale ?

**Comment by devolution79** [comment]
I need to be able to generate random maps with large rivers. Alternatively, somebody could supply me with test-maps that are known to be problematic!

**Comment by devolution79** [comment]
A save would be helpful too, but I think that is problematic since New_Hope is undergoing rapid development and the save may not be compatible (unless we agree on which revision was used)

**Comment by raystuttgart** [comment]
@devolution79 
Then maybe just ask **Ramstormp** to give you the MapScript he had merged / created from your MapScript and the MapScript of Flavius Belisarius.

Alternatively you could simply use one of the **existing Maps** to test.
All of the have **Large Rivers** already and most have already been adjusted to New Hope.

What is the difference if the Map was created by a MapScript or by a Map maker / person?
For your tests it should hardly matter if it was randomly generated or made by hand.

**Comment by devolution79** [comment]
I've committed the new land area generator. Please help test this change!

**Comment by raystuttgart** [comment]
Awesome !

**Comment by MrZorG33** [comment]
I tested the fix on the Two Continents map (now 1690). despite the abundance of LRs and river fords, Scouts and automated units move without problems.
